### PR TITLE
Remove copy/move constructor of SpinMutex, and align thread caches to cache line

### PR DESCRIPTION
**Pull request by JiayuZzz**
Signed-off-by: Jiayu Wu <jiayu.wu@intel.com>

### What problem does this PR solve?

Problem Summary:

- SpinMutex has copy and move constructor, which is insecure
- Thread caches are not aligned to cache line size, so they may need to do cache coherency which will decrease performance

### What is changed and how it works?

What's Changed:
- Remove copy/move constructor and assignment of SpinMutex for security
- Implement Array to reduce thread cache size, and store un-copyable elements
- align thread cache size to 64B

### Check List 

Tests 

- Unit test
- Stress test

Side effects

- 1% ~ 3% improvement on string insert
- no significant difference on other types

### CODE REVIEW

File: engine/utils.hpp [hunk 1]
```diff
@@ -117,6 +117,41 @@ static inline bool equal_string_view(const StringView &src,
   return false;
 }
 
+template <typename T> class FixedVector {
+public:
+  FixedVector(uint64_t size) : size_(size) { data_ = new T[size]; }
```
Comment (ZiyanShi) [inline engine/utils.hpp:122]: explicit

File: engine/utils.hpp [hunk 2]
```diff
@@ -117,6 +117,41 @@ static inline bool equal_string_view(const StringView &src,
...
+
+  FixedVector() = delete;
+
+  FixedVector(const FixedVector<T> &v) {
+    size_ = v.size_;
+    data_ = new T[size_];
+    memcpy(data_, v.data_, size_ * sizeof(T));
```
Comment (ZiyanShi) [inline engine/utils.hpp:129]: Please don't.
If you have to, use SFINAE to limit T as POD.

File: engine/utils.hpp [hunk 3]
```diff
@@ -117,6 +117,41 @@ static inline bool equal_string_view(const StringView &src,
...
+  }
+
+  ~FixedVector() {
+    if (data_ != nullptr) {
+      delete[] data_;
+    }
```
Comment (ZiyanShi) [inline engine/utils.hpp:135]: Explicitly call destructor ~T().
Otherwise resources hold by T will be leaked.

File: engine/utils.hpp (hunk 1 above)
Comment (ZiyanShi) [reply engine/utils.hpp:122]: No valid object is constructed.
Assigning to reference got from operator[] is UB.

File: engine/pmem_allocator/free_list.hpp [hunk 4]
```diff
@@ -190,14 +190,17 @@ class Freelist {
           last_used_entry_ts(0) {}
 
     // Entry size stored in block unit
-    std::vector<std::vector<SpaceEntry>> active_entries;
+    FixedVector<std::vector<SpaceEntry>> active_entries;
```
Comment (ZiyanShi) [inline engine/pmem_allocator/free_list.hpp:193]: Why not use std::array or std::vector initialized to certain size?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/pmem_allocator/free_list.hpp (hunk 4 above)
Comment (JiayuZzz) [reply engine/pmem_allocator/free_list.hpp:193]: we need variable size

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**